br/pkg/lightning/backend/external/concurrentReader.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/util/mathutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"io"
+)
+
+// SingeFileReader is a concurrent reader for a single file.
+type SingeFileReader struct {
+	ctx               context.Context
+	concurrency       int
+	readBufferSize    int
+	currentFileOffset int64
+	bufferReadOffset  int64
+	bufferMaxOffset   int64
+
+	BufferTotalOffset int64
+
+	MaxFileOffset int64
+	name          string
+
+	storage storage.ExternalStorage
+	buffer  []byte
+}
+
+// NewSingeFileReader creates a new SingeFileReader.
+func NewSingeFileReader(ctx context.Context, st storage.ExternalStorage, name string, concurrency int, readBufferSize int) (*SingeFileReader, error) {
+	if st == nil {
+		return nil, nil
```

Comment:
maybe we can return error or panic in these unexpected usage

Reply:
I don't return error so that it can work for the current tests.